Write assets to a temporary directory before calling out to sass

nshahan opened the pull request:
* Copy primary assets and all transitive imports to a temporary directory. Build css using the copied sources. Ensures that all assets are available on disk before calling out to the sass compiler.
* Add a pub transformer that wraps `SassBuilder`.
* Add tests to validate parsing of imports.

nshahan committed:
Write assets to a temporary directory before calling out to sass

Copy primary assets and all transitive imports to a temporary directory. Build css using the copied sources. Ensures that all assets are available on disk before calling out to the sass compiler.
Add a pub transformer that wraps `SassBuilder`.
Add tests to validate parsing of imports.

nshahan commented:
Implements the temp directory workaround @jakemac53 listed in #1.

jakemac53 commented on lib/sass_builder.dart:
```diff
@@ -1,94 +1,148 @@
-import 'dart:io';
+import 'dart:collection';
+
+import 'package:barback/barback.dart' show BarbackSettings;
```
Ideally, this file wouldn't import package:barback. You could change the constructor to just take a file extension instead and deal with parsing this settings object inside the `transformer.dart` file.

jakemac53 commented on lib/sass_builder.dart:
```diff
@@ -1,94 +1,148 @@
+    // Read and copy this asset and all imported assets to the temp directory.
+    _log.fine('processing file: ${inputId}');
+    var tempDir =
+        await buildStep.fetchResource<ScratchSpace>(_scratchSpaceResource);
```
nit: you shouldn't need the explicit generic type here (it should get inferred)

jakemac53 commented on lib/sass_builder.dart:
```diff
@@ -1,94 +1,148 @@
+    assetsToCopy.add(id);
+
+    while (assetsToCopy.isNotEmpty) {
+      id = assetsToCopy.removeFirst();
```
I think  you also need a `Set<AssetId>` of previously processed assets, and you could either `continue` here if `id` is already in that or make sure you don't add ones you have already seen on line 93 below.

nshahan replied:
I had that check in my original version as a guard against copying a file twice but I removed it since the ScratchSpace now provides the guarantee for me. 

This could get stuck running forever if you gave it sass files with a cycle of imports. I'll add the check back in.

jakemac53 commented on lib/sass_builder.dart:
```diff
@@ -1,94 +1,148 @@
+        var importId = new AssetId(_importPackage(import.group(1), id.package),
+            _importPath(import.group(1), id.path));
+
+        if (!await (buildStep.canRead(importId))) {
```
nit: unnecessary parenthesis

jakemac53 commented on lib/sass_builder.dart:
```diff
@@ -1,94 +1,148 @@
+              join(dirname(importId.path), '_${basename(importId.path)}'));
+        }
+
+        if (!await buildStep.canRead(importId)) {
```
I would move this logic inside the if above, otherwise you are often checking the same import twice.

jakemac53 commented on sass_builder_test.dart:
```diff
@@ -0,0 +1,207 @@
+
+      reader.cacheStringAsset(primary, inputs[primary]);
+
+      await runBuilder(builder, inputs.keys, reader, writer, const BarbackResolvers());
```
nit: you aren't resolving any code so i think you could pass `null` for the last arg, and not have a dependency on build_barback in this file.

jakemac53 commented on sass_builder_test.dart:
```diff
@@ -0,0 +1,207 @@
+      expect(writer.assets, contains(primary.changeExtension('.css')));
+
+      expect(reader.assets.length, equals(2));
+      expect(reader.assetsRead, contains(primary));
```
nit: `allOf(contains(primary), contains(import))` or `unorderedEquals([primary, import])` read a bit better to me (same below)

jakemac53 reviewed (commented):
looks good!

nshahan committed:
Changes for review

Remove barback dependencies where possible.
Add check for assets that have already been read and copied.
Move sass_builder_test.dart to test directory.
Update expected behavior in tests to use iterable matchers when possible.

jakemac53 commented:
This LGTM from a package:build perspective

luisvt commented:
sorry, I should revert the merge doubt to next error when trying to run `tool/build.dart`:

```
[SEVERE] Build: Failed after 413ms
UnexpectedExistingOutputsException: Either you opted not to delete existing files, or you are not running in interactive mode and did not specify `deleteFilesByDefault` as `true`.
package:build_runner/src/generate/build_impl.dart 316:7   BuildImpl._deletePreviousOutputs
package:build_runner/src/generate/build_impl.dart 149:15  BuildImpl._safeBuild.<fn>.<fn>
package:build_runner/src/logging/logging.dart 19:28       logWithTime
package:build_runner/src/generate/build_impl.dart 133:13  BuildImpl._safeBuild.<fn>
package:stack_trace                                       Chain.capture
package:build_runner/src/generate/build_impl.dart 88:11   BuildImpl._safeBuild
package:build_runner/src/generate/build_impl.dart 64:24   BuildImpl.runBuild
package:build_runner/src/generate/build.dart 62:32        build
tool/build.dart 7:9                                       main
dart:async                                                _Completer.completeError
package:build_runner/src/generate/build_impl.dart 344:3   BuildImpl._deletePreviousOutputs
dart:async                                                new Future.microtask
package:build_runner/src/generate/build_impl.dart 284:68  BuildImpl._deletePreviousOutputs
package:build_runner/src/generate/build_impl.dart 149:15  BuildImpl._safeBuild.<fn>.<fn>
dart:async                                                _asyncThenWrapperHelper
package:build_runner/src/generate/build_impl.dart 133:69  BuildImpl._safeBuild.<fn>.<fn>
package:build_runner/src/logging/logging.dart 19:28       logWithTime
dart:async                                                new Future.microtask
package:build_runner/src/logging/logging.dart 16:66       logWithTime
package:build_runner/src/generate/build_impl.dart 133:13  BuildImpl._safeBuild.<fn>
dart:async                                                _asyncThenWrapperHelper
package:build_runner/src/generate/build_impl.dart 88:28   BuildImpl._safeBuild.<fn>
package:stack_trace                                       Chain.capture
package:build_runner/src/generate/build_impl.dart 88:11   BuildImpl._safeBuild
package:build_runner/src/generate/build_impl.dart 64:24   BuildImpl.runBuild
```

jakemac53 commented:
I don't think that error is really a problem, but just a bad interaction
with the previous behavior. Since it wasn't using the `BuildStep` to write
files before package:build did not know about the outputs, so the first
time you run it with the new behavior it will complain about the files it
wants to generate already existing (by default it won't overwrite what it
thinks are source files).

If you run the script directly from a console it should prompt you to
delete the pre-existing files and you should just be able to type `y` and
approve that, or you can add `deleteFilesByDefault: true` to your
invocation of build/watch/serve.

On Wed, Sep 27, 2017 at 8:34 PM Luis Vargas <notifications@github.com>
wrote:

> sorry, I should revert the merge doubt to next error when trying to run
> tool/build.dart:
>
> [SEVERE] Build: Failed after 413ms
> UnexpectedExistingOutputsException: Either you opted not to delete existing files, or you are not running in interactive mode and did not specify `deleteFilesByDefault` as `true`.
> package:build_runner/src/generate/build_impl.dart 316:7   BuildImpl._deletePreviousOutputs
> package:build_runner/src/generate/build_impl.dart 149:15  BuildImpl._safeBuild.<fn>.<fn>
> package:build_runner/src/logging/logging.dart 19:28       logWithTime
> package:build_runner/src/generate/build_impl.dart 133:13  BuildImpl._safeBuild.<fn>
> package:stack_trace                                       Chain.capture
> package:build_runner/src/generate/build_impl.dart 88:11   BuildImpl._safeBuild
> package:build_runner/src/generate/build_impl.dart 64:24   BuildImpl.runBuild
> package:build_runner/src/generate/build.dart 62:32        build
> tool/build.dart 7:9                                       main
> dart:async                                                _Completer.completeError
> package:build_runner/src/generate/build_impl.dart 344:3   BuildImpl._deletePreviousOutputs
> dart:async                                                new Future.microtask
> package:build_runner/src/generate/build_impl.dart 284:68  BuildImpl._deletePreviousOutputs
> package:build_runner/src/generate/build_impl.dart 149:15  BuildImpl._safeBuild.<fn>.<fn>
> dart:async                                                _asyncThenWrapperHelper
> package:build_runner/src/generate/build_impl.dart 133:69  BuildImpl._safeBuild.<fn>.<fn>
> package:build_runner/src/logging/logging.dart 19:28       logWithTime
> dart:async                                                new Future.microtask
> package:build_runner/src/logging/logging.dart 16:66       logWithTime
> package:build_runner/src/generate/build_impl.dart 133:13  BuildImpl._safeBuild.<fn>
> dart:async                                                _asyncThenWrapperHelper
> package:build_runner/src/generate/build_impl.dart 88:28   BuildImpl._safeBuild.<fn>
> package:stack_trace                                       Chain.capture
> package:build_runner/src/generate/build_impl.dart 88:11   BuildImpl._safeBuild
> package:build_runner/src/generate/build_impl.dart 64:24   BuildImpl.runBuild
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dart-league/sass_builder/pull/7#issuecomment-332718293>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA8HWS3eXx0dJGIx2Wd5HypDdksxbZPfks5smxO4gaJpZM4PhGVg>
> .
>
